### DISCUSSION TITLE
EntityRepository not persisting Entities when using `find` and flush after property update.

### DISCUSSION THREAD

**Discussion by mboudreau** [mikro-orm/mikro-orm#5833] [General]
### Describe the bug

This has been bugging me for months, but I didn't have time to actually figure out what's going on because we needed to deliver now and found that using `persistAndFlush` seems to work, but I know it's not the proper way of doing it and would like to know what's going on.

The gist of it is that most of our Entities use the Repository pattern because there's a lot of logic for our entities and it makes it much easier to maintain.  A simple example that I had an issue with recently:  I fork the EntityManager (like I do for every request), then use the EntityRepository to get a particular Entity (or multiple), then I would update said entity with a changed property, then would call `await em.flush()` but the new property changed on the Entity would never get updated in the database.  The only way that the property actually get updated in the DB is to forcefully `persistAndFlush(entity)`.

I cannot seem to find anything in the documentation around this.  I'm guessing there's some kind of loss of communication between the forked EM and the global EM that the EntityRepository uses?  But at the same time, I cannot find a way to pass in the EM to the Repository.

Please help, this has caused multiple issues for us thinking of simple change will persist, and then it just doesn't...

### Reproduction

```
const orm = MikroORM.initSync<PostgreSqlDriver>(config);
        
@Entity({ tableName: "user", repository: () => UserRepository })
export class User {
	[EntityRepositoryType]?: UserRepository;
	
       @Property()
	isExpired: boolean = false

       @Property({ nullable: true })
        expires_at: Date
}

export class UserRepository extends EntityRepository<User> {
    public async findExpired() {
	return this.findAll({where: { expires_at: { $lte: new Date() } } });
    }
}

export const example = async () => {
	const em = orm.em.fork();
	const expiredUsers = await em.getRepository(User).findExpired();
	expiredUsers.forEach(user => {
		user.isExpired = true;
	});
	await em.flush();
};
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2.9

### Node.js version

18.18.2

### Operating system

Ubuntu

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> Reproduction

Is that really a failing repro or a snippet you haven't tried executing? I am rather sure that if I convert this one into an executable one, it will work as expected.

What you described above sounds like you have a context mismatch. You could see that if you compare `em.id` of the EM behind the repository and the one the managed entity is holding. The fact that you say using `persistAndFlush` helps kinda confirms this, since it means the entity was not managed by the EM you flush before.

You can get the EM id of a managed entity like this:

```ts
console.log(wrap(entity, true).__em.id);
```

>  I cannot find a way to pass in the EM to the Repository.

You don't pass it to the repository, you create the repository from the EM (via `em.getRepository`).

**Reply by B4nan** [reply]
Do you fork explicitly in your app (like in this repro)? Are you using the `CreateRequestContext` decorator by any chance? There is a bug ([already fixed in master](https://github.com/mikro-orm/mikro-orm/commit/1bb4e22e17537600ef7f10f02aa638c3b9ce609e)) in there, it actually behaves like `EnsureRequestContext` now, respecting the existing context, which could also result in weird behavior like this one.
